### ISSUE TITLE
Lost files if more than 25 files.

### ISSUE THREAD

**Issue by judewang** [jonkemp/gulp-useref#102]
Hello, I'm using [generator-gulp-webapp](https://github.com/yeoman/generator-gulp-webapp) and have 27 html files compiled to the `.tmp` folder. I found only the first 25 html files will be process and pass to the `dist` folder during the gulp-useref task. Here's the `gulpfile.coffee`'s code, really need to know how to fix this, thanks.

```
gulp.task 'views', ->
  gulp.src [
      '!app/mixins/**/*.jade'
      '!app/layouts/**/*.jade'
      '!app/partials/**/*.jade'
      'app/**/*.jade'
    ], base: 'app'
    .pipe $.data (file) ->
      return JSON.parse fs.readFileSync '.tmp/data.json'
    .pipe $.jade
      pretty: true
      basedir: 'app/'
      compileDebug: true
    .on 'error', (err) ->
      $.notify
        title: 'Jade compile failed.',
        sound: 'Sosumi'
      .write err.message
    .pipe gulp.dest '.tmp'

gulp.task 'htmlProcess', ['views', 'styles', 'scripts', 'compass'], ->
  assets = $.useref.assets
    searchPath: ['.tmp', 'app', '.']

  gulp.src ['app/**/*.html', '.tmp/**/*.html']
    .pipe assets
    .pipe assets.restore()
    .pipe $.useref()
    .pipe gulp.dest 'dist'
```

**Comment by foria** [comment]
+1 for me more then 17

**Comment by dheerajvs** [comment]
+1 for me it stop after ~80 files. This is a serious bug.

**Comment by simonswiss** [comment]
+1 after 27 files, running the gulp task in isolation.

**Comment by mingjunyang** [comment]
无知的我整了372 个html文件，现在已经哭晕在厕所了。有一个奇怪的现象就是，每次丢失的都是相同的文件。在openSUSE13.2执行几次命令之后有可能不丢失，但是在ubuntu 14.04.2 LTS 则总是丢失文件。
以下我来书写英文描述，比较难懂。可以猜猜中文意思吧。

I use `gulp` package my  web,and I use the `gulp-useref`.

```javascript
var useref = require('gulp-useref')
# require others

gulp.task('html', function() {
    var assets = useref.assets({
        searchPath: ['app']
    });

    return gulp.src('app/**/*.html')
        .pipe(assets)
        .pipe(gif('*.js', uglify({
            mangle: false,
        })))
        .pipe(gif('*.css', minifyCss({
            compatibility: '*'
        })))
        .pipe(gif('*.html', minifyHtml({
            conditionals: true,
            loose: true
        })))
        .pipe(assets.restore())
        .pipe(useref())
        .pipe(gulp.dest('dist'));
});

```
on my openSUSE13.2 sometime I run `gulp html` more than 3 times, maybe it's okay. But on ubuntu14.04,lost files any times.
And  every time lost is the same files.
Oh, lost HTML files.

**Comment by judewang** [comment]
@mingjunyang 你改用手動 copy html 到 dist 資料夾會把引用 js 跟 css 的路徑原封不動 copy 過去吧？那個路徑的 js 跟 css 是尚未 concat 跟 minify 的，這樣應該還是沒有解決問題吧？

**Comment by mingjunyang** [comment]
@judewang 'resourceCopy'命令只针对template文件夹下的*.html文件。我这里写的是AngularJS的single page，只有index.html是引入js和css文件的。其它的html files只需要原样拷贝过来就好了。本来是想偷个懒，用useref把那些template files拷贝过去。

**Comment by dahart** [comment]
I bumped into this issue yesterday. Not sure if its getting any love any time soon, but I did manage to put together a temporary workaround. You can split your task into multiple parts, and use gulp-cached to prevent overwrites.


```
var autoprefixer = require('gulp-autoprefixer');
var cache        = require('gulp-cached');
var csso         = require('gulp-csso');
var filter       = require('gulp-filter');
var useref       = require('gulp-useref');

...

var makeCssBuildTask = function(taskName, files, taskDeps) {
  gulp.task(taskName, 'Build minified css files into dest', taskDeps, 
    function () {
      return gulp.src([files])
        .pipe(useref.assets())                // Replace the stream with the files referenced in <build> tags
        .pipe(cache('build.css'))             // Remove dupes, string == cache name
        .pipe(filter('**/*.css'))             // Filter out everything except the css
        .pipe(autoprefixer('last 1 version')) // Auto-prefix css tags that have browser specific implementations
        .pipe(csso())                         // Minify the css
        .pipe(gulp.dest(destDir));
    });
}

makeCssBuildTask("build.1.css", "[a-m]*.html", []);
makeCssBuildTask("build.2.css", "[n-z]*.html", ["build.1.css"]);

gulp.task('build.css', 'Build minified css files into dest', ["build.2.css"]);
```

I made my build depend on build.css, which then executes build.1.css followed by build.2.css. When I read *.html, my stream fails to build all css files, but when I break it into two parts like this, everything works.

I also have a note to myself that putting html files in the stream that do not have a 'build:css' tag in them can cause the stream to fail. Maybe some people are having that problem?

**Comment by jonkemp** [comment]
Can we make a failing test for this?

**Comment by dahart** [comment]
Okay, here's a fairly minimal failing test.

makeUserefTest:
```
#!/bin/bash

for f in $(seq -w 39); do
	cat <<- END > useref.$f.html
		<html>
			<head>
				<!-- build:css useref.$f.min.css -->
				<link rel="stylesheet" type="text/css" href="useref.$f.css"/>
				<!-- endbuild -->
			</head>
			<body></body>
		</html>
	END

	touch useref.$f.css

done
```

gulpfile.js:
```
'use strict';

var gulp   = require('gulp-help')(require('gulp'));
var useref = require('gulp-useref');

var makeCssBuildTask = function(taskName, desc, files, taskDeps) {
  gulp.task(taskName, desc, taskDeps,
    function () {
      return gulp.src([files])
        .pipe(useref.assets())                // Replace the stream with the files referenced in <build> tags
        .pipe(gulp.dest('dest'));
    });
}

makeCssBuildTask('working.1', null, 'useref.0*.html', []);
makeCssBuildTask('working.2', null, 'useref.1*.html', ['working.1']);
makeCssBuildTask('working.3', null, 'useref.2*.html', ['working.2']);
makeCssBuildTask('working.4', null, 'useref.3*.html', ['working.3']);
makeCssBuildTask('working.5', null, 'useref.4*.html', ['working.4']);
gulp.task('working', 'This test succeeds', ['working.4']);

makeCssBuildTask('broken', 'This test fails', 'useref.*.html', []);
```

sample run output:
```
~/tmp
> . makeUserefTest

~/tmp
> gulp broken
[19:37:23] Using gulpfile ~/tmp/gulpfile.js
[19:37:23] Starting 'broken'...

~/tmp
> ls dest | wc
     17      17     306

~/tmp
> gulp working
[19:37:41] Using gulpfile ~/tmp/gulpfile.js
[19:37:41] Starting 'working.1'...
[19:37:41] Finished 'working.1' after 213 ms
[19:37:41] Starting 'working.2'...
[19:37:41] Finished 'working.2' after 36 ms
[19:37:41] Starting 'working.3'...
[19:37:41] Finished 'working.3' after 36 ms
[19:37:41] Starting 'working.4'...
[19:37:41] Finished 'working.4' after 24 ms
[19:37:41] Starting 'working'...
[19:37:41] Finished 'working' after 7.06 μs

~/tmp
> ls dest | wc
     39      39     702
```

**Comment by Kikketer** [comment]
Running into the same issue here.  This is a serious issue and needs to be rectified since "usemin" is blacklisted by gulp.  I have about 80 files and a good portion of them are getting cut out (which makes for strange errors when concatinated since random parts of the app break).

**Comment by jonkemp** [comment]
I am currently working on version 2.0 which is simply a refactor, no new features are being added. I am hopeful this will fix some of the issues people are experiencing. 

I am aware of this issue but I am not sure how to solve it other than refactoring. Help is appreciated. This work is being done in the 'refactor' branch if you want to test it against your project. There is a workaround posted above in the meantime. Thanks.

**Comment by jonkemp** [comment]
Thanks for the failing test case @dahart. I think I found and fixed the issue. Checkout the latest version of this repo and link it or wait for the next version to be published on npm.

**Comment by jonkemp** [comment]
Fixed in 2.0.0.

79c2a0e
